src/main/java/org/kohsuke/github/GitHubRequest.java

```diff
@@ -0,0 +1,651 @@
+package org.kohsuke.github;
+
+import org.apache.commons.lang3.StringUtils;
+
+import java.io.InputStream;
+import java.io.UnsupportedEncodingException;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Consumer;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import javax.annotation.WillClose;
+
+import static java.util.Arrays.asList;
+
+/**
+ * Class {@link GitHubRequest} represents an immutable instance used by the client to determine what information to
+ * retrieve from a GitHub server. Use the {@link Builder} construct a {@link GitHubRequest}.
+ * <p>
+ * NOTE: {@link GitHubRequest} should include the data type to be returned. Any use cases where the same request should
+ * be used to return different types of data could be handled in some other way. However, the return type is currently
+ * not specified until late in the building process, so this is still untyped.
+ * </p>
+ */
+class GitHubRequest {
+
+    private static final List<String> METHODS_WITHOUT_BODY = asList("GET", "DELETE");
+    private final List<Entry> args;
+    private final Map<String, String> headers;
+    private final String apiUrl;
+    private final String urlPath;
+    private final String method;
+    private final InputStream body;
+    private final boolean forceBody;
+
+    private final URL url;
+
+    private GitHubRequest(@Nonnull List<Entry> args,
+            @Nonnull Map<String, String> headers,
+            @Nonnull String apiUrl,
+            @Nonnull String urlPath,
+            @Nonnull String method,
+            @CheckForNull InputStream body,
+            boolean forceBody) throws MalformedURLException {
+        this.args = Collections.unmodifiableList(new ArrayList<>(args));
+        this.headers = Collections.unmodifiableMap(new LinkedHashMap<>(headers));
+        this.apiUrl = apiUrl;
+        this.urlPath = urlPath;
+        this.method = method;
+        this.body = body;
+        this.forceBody = forceBody;
+        String tailApiUrl = buildTailApiUrl();
+        url = getApiURL(apiUrl, tailApiUrl);
+    }
+
+    /**
+     * Create a new {@link Builder}.
+     * 
+     * @return a new {@link Builder}.
+     */
+    public static Builder<?> newBuilder() {
+        return new Builder<>();
+    }
+
+    /**
+     * Gets the final GitHub API URL.
+     */
+    @Nonnull
+    static URL getApiURL(String apiUrl, String tailApiUrl) throws MalformedURLException {
+        if (tailApiUrl.startsWith("/")) {
```

Comment:
Is there a circumstance where `tailApiUrl` will _not_ start with a "/"? I didn't see any tests that behave this way, so just wondering how we'd run up against this, and what it looks like when we do.

Reply:
Good question.  There are a small number of cases where the API returns a URL that is then used for other things.  

https://github.com/github-api/github-api/pull/697/files#diff-64dfb4dfbe21520e8735454a113c45a7R548-R560


https://github.com/github-api/github-api/blob/ff3136df7069d22d937daa8f2cddc357c777e07e/src/main/java/org/kohsuke/github/GHContent.java#L225-L227

https://github.com/github-api/github-api/blob/ff3136df7069d22d937daa8f2cddc357c777e07e/src/main/java/org/kohsuke/github/GHContent.java#L236-L240

Also, for paginated result, the full url for the next page is provided in the `Link` header , so when we call `getApiURL` we want to use that url, whatever it happens to be.

https://github.com/github-api/github-api/blob/dc33e28452532b94c6d46e435c192ae043849ca9/src/main/java/org/kohsuke/github/GitHubPageResponseIterator.java#L68-L77

I don't love this lack of clarity, but I didn't feel confident that I had tests for all cases where that URL didn't start with `/`, so I've kept the behavior where this method just handles it.  That said, I should change L77 to use `setRawUrl`.  Thanks!